compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -1911,14 +1912,23 @@ public void visit(BLangAnnotAccessExpr annotAccessExpr) {
 
     @Override
     public void visit(BLangStatementExpression bLangStatementExpression) {
-        if (bLangStatementExpression.stmt.getKind() == NodeKind.BLOCK) {
-            BLangBlockStmt bLangBlockStmt = (BLangBlockStmt) bLangStatementExpression.stmt;
+        BLangStatement exprStmt = bLangStatementExpression.stmt;
+        boolean isBlockStmt = exprStmt.getKind() == NodeKind.BLOCK;
+        if (isBlockStmt && !((BLangBlockStmt) exprStmt).isLetExpr) {
+            BLangBlockStmt bLangBlockStmt = (BLangBlockStmt) exprStmt;
             for (int i = 0; i < bLangBlockStmt.stmts.size(); i++) {
                 BLangStatement stmt = bLangBlockStmt.stmts.remove(i);
                 bLangBlockStmt.stmts.add(i, rewrite(stmt, env));
             }
         } else {
-            bLangStatementExpression.stmt = rewrite(bLangStatementExpression.stmt, env);
+            // In let expression's object constructor expression, the function block already has an OCE mapSymbol,
+            // which should be propagated to block statement's mapSymbol.
+            if (isBlockStmt && bLangStatementExpression.getBType().tag == TypeTags.OBJECT &&
+                    ((BLangBlockStmt) exprStmt).mapSymbol == null &&
+                    env.node.getKind() == NodeKind.BLOCK_FUNCTION_BODY) {
+                ((BLangBlockStmt) exprStmt).mapSymbol = ((BLangBlockFunctionBody) env.node).mapSymbol;
```

Comment:
Now that, we anyway have a nested if condition, we can move this logic inside the 
`if (bLangStatementExpression.stmt.getKind() == NodeKind.BLOCK) {}`, while keeping the else block intact.